pageserver/src/tenant.rs

```diff
@@ -782,27 +810,74 @@ impl Tenant {
             "initial tenant load",
             false,
             async move {
+                let make_broken = |t: &Tenant, err: anyhow::Error| {
+                    error!("load failed, setting tenant state to Broken: {err:?}");
+                    t.state.send_modify(|state| {
+                        assert_eq!(
+                            *state,
+                            TenantState::Loading,
+                            "the loading task owns the tenant state until activation is complete"
+                        );
+                        *state = TenantState::broken_from_reason(err.to_string());
+                    });
+                };
+
                 let mut init_order = init_order;
 
                 // take the completion because initial tenant loading will complete when all of
                 // these tasks complete.
-                let _completion = init_order.as_mut().and_then(|x| x.initial_tenant_load.take());
+                let _completion = init_order
+                    .as_mut()
+                    .and_then(|x| x.initial_tenant_load.take());
+
+                // Dont block pageserver startup on figuring out deletion status
+                let pending_deletion = {
+                    match DeleteTenantFlow::should_resume_deletion(
+                        conf,
+                        remote_storage.as_ref(),
+                        &tenant_clone,
+                    )
+                    .await
+                    {
+                        Ok(should_resume_deletion) => should_resume_deletion,
+                        Err(err) => {
+                            make_broken(&tenant_clone, anyhow::anyhow!(err));
+                            return Ok(());
+                        }
+                    }
+                };
+
+                let background_jobs_can_start =
+                    init_order.as_ref().map(|x| &x.background_jobs_can_start);
+
+                if let Some(deletion) = pending_deletion {
+                    match DeleteTenantFlow::resume(
```

Comment:
```suggestion
                if let Some(deletion) = pending_deletion {
                    // as we are no longer loading, signal completion by dropping
                    // the completion while we resume deletion
                    drop(_completion);
                    match DeleteTenantFlow::resume(
```

Otherwise it should hang on startup, because init_done_rx never completes in `pageserver/src/main.rs`, if the `background_jobs_can_start` are used within this path.

Reply:
the barriers are an obvious deadlock hazard... should create a crate which would panic if you wait on a barrier while keeping up a guard. Problem is of course how to do this transitively...

Reply:
Good catch!

Reply:
Done https://github.com/neondatabase/neon/pull/4855/commits/5a42259332353cb6506fa0796bc90548613a80a8